olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp

```diff
@@ -40,20 +43,63 @@ VolatileLayerClientImpl::VolatileLayerClientImpl(
     client::OlpClientSettings settings)
     : catalog_(std::move(catalog)),
       layer_id_(std::move(layer_id)),
-      settings_(std::move(settings)),
+      settings_(
+          std::make_shared<client::OlpClientSettings>(std::move(settings))),
       pending_requests_(std::make_shared<PendingRequests>()) {
-  if (!settings_.cache) {
-    settings_.cache = client::OlpClientSettingsFactory::CreateDefaultCache({});
+  if (!settings_->cache) {
+    settings_->cache = client::OlpClientSettingsFactory::CreateDefaultCache({});
   }
   // to avoid capturing task scheduler inside a task, we need a copy of settings
   // without the scheduler
-  task_scheduler_ = std::move(settings_.task_scheduler);
+  task_scheduler_ = std::move(settings_->task_scheduler);
+
+  auto cache = settings_->cache;
+
+  auto api_repo =
+      std::make_shared<repository::ApiRepository>(catalog_, settings_, cache);
+
+  catalog_repo_ = std::make_shared<repository::CatalogRepository>(
+      catalog_, api_repo, cache);
+
+  partition_repo_ = std::make_shared<repository::PartitionsRepository>(
+      catalog_, api_repo, catalog_repo_, cache);
 }
 
 VolatileLayerClientImpl::~VolatileLayerClientImpl() {
   pending_requests_->CancelPendingRequests();
 }
 
+client::CancellationToken VolatileLayerClientImpl::GetPartitions(
+    PartitionsRequest request, Callback<PartitionsResponse> callback) {
+  request.WithLayerId(layer_id_);
+  client::CancellationToken token;
+  int64_t request_key = pending_requests_->GenerateRequestPlaceholder();
+  auto pending_requests = pending_requests_;
+  auto request_callback = [pending_requests, request_key,
+                           callback](PartitionsResponse response) {
+    if (pending_requests->Remove(request_key)) {
+      callback(response);
+    }
+  };
+  if (CacheWithUpdate == request.GetFetchOption()) {
+    auto req = request;
+    token = partition_repo_->GetPartitions(req.WithFetchOption(CacheOnly),
+                                           request_callback);
+    auto onlineKey = pending_requests_->GenerateRequestPlaceholder();
+    pending_requests_->Insert(
+        partition_repo_->GetPartitions(
+            req.WithFetchOption(OnlineIfNotFound),
+            [pending_requests, onlineKey](PartitionsResponse) {
+              pending_requests->Remove(onlineKey);
```

Comment:
Should we call the callback here? 
```suggestion
              pending_requests->Remove(onlineKey);
              request_callback(callback);
```

Reply:
No, that way callback will be called twice: first time on CacheOnly operation (at 86-87) and second time on Online